# Đồ án thực hành: XÂY DỰNG VÀ KHAI THÁC KDL
- Môn: **HTTT Phục vụ trí tuệ kinh doanh**
- Lớp: **19HTTT2**
- Yêu cầu: **Xây dựng và phân tích dữ liệu về Covid-19 trong các năm 2020 - 2022**
- Thông tin nhóm:

|STT | MSSV | Họ tên |
| :---: | :---: | :--- |
| 1 | 19127304 | Trần Khải Trúc |
| 2 | 19127474 | Lương Kiếm Minh | 
| 3 | 19127649 | Tô Thanh Tuấn | 

In [7]:
import pandas as pd
import string

## 1. Chuẩn bị dữ liệu `Cases Report`

### Đọc và xem bảng

In [8]:
# Đọc dữ liệu
f_path = 'Datasource/Cases Report.csv'
case_report_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = case_report_df.columns.str.title()
case_report_df.columns = [header.replace(' ','_').upper() for header in header_list]

# Xem thử dữ liệu
case_report_df.head()

OUTCOME  AGE  GENDER                      REPORTING_PHU SPECIMENDATE  \
0  Resolved  80s  FEMALE              Toronto Public Health    9/24/2020   
1  Resolved  <20    MALE                 Peel Public Health   12/24/2020   
2  Resolved  20s    MALE              Lambton Public Health    9/23/2020   
3  Resolved  20s    MALE  Region of Waterloo, Public Health    9/20/2020   
4  Resolved  20s  FEMALE               Ottawa Public Health    9/21/2020   

  CASEREPORTED_DATE       PHUCITY TESTREPORTED_DATE CASEACQUISITION_INFO  \
0          2/6/2021       Toronto          2/6/2021                   OB   
1          2/1/2021   Mississauga          2/1/2021                   CC   
2         1/15/2021  Point Edward         1/14/2021                   CC   
3        12/29/2020      Waterloo         9/23/2020    NO KNOWN EPI LINK   
4        12/24/2020        Ottawa         9/22/2020                   CC   

  ACCURATEEPISODE_DT                     PHU_ADDRESS  \
0          1/26/2021  277 Victoria Street, 5th Floor   
1          1/29/2021          7120 Hurontario Street   
2          1/10/2021              160 Exmouth Street   
3          9/20/2020          99 Regina Street South   
4          9/21/2020         100 Constellation Drive   

                                         PHU_WEBSITE OUTBREAKRELATED  \
0  www.toronto.ca/community-people/health-wellnes...             Yes   
1                          www.peelregion.ca/health/             NaN   
2                            www.lambtonhealth.on.ca             Yes   
3                            www.regionofwaterloo.ca             NaN   
4                          www.ottawapublichealth.ca             NaN   

   PHU_LATITUDE  PHU_LONGITUDE PHU_POSTAL_CODE  
0     43.656591     -79.379358         M5B 1W2  
1     43.647471     -79.708893         L5W 1N4  
2     42.986416     -82.404808         N7T 7Z6  
3     43.462876     -80.520913         N2J 4V3  
4     45.345665     -75.763912         K2G 6J8

### Matching data

In [9]:
# Đổi tên một số cột theo bảng mô tả dữ liệu của để bài
case_report_df.rename(columns = {'AGE':'AGE_GROUP', 'CASEREPORTED_DATE':'REPORTED_DATE', 'PHUCITY':'REPORTING_PHU_CITY', 'PHU_ADDRESS':'REPORTING_PHU_ADDRESS', 'PHU_WEBSITE':'REPORTING_PHU_WEBSITE', 'PHU_LATITUDE':'REPORTING_PHU_LATITUDE', 'PHU_LONGITUDE':'REPORTING_PHU_LONGITUDE', 'PHU_POSTAL_CODE':'REPORTING_PHU_POSTAL_CODE'}, inplace = True)

# Đổi value của một số cột để khớp với các bảng
# Age_Group
case_report_df.loc[case_report_df["AGE_GROUP"] == "<20", "AGE_GROUP"] = "<20yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "20s", "AGE_GROUP"] = "20-29yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "30s", "AGE_GROUP"] = "30-39yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "40s", "AGE_GROUP"] = "40-49yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "50s", "AGE_GROUP"] = "50-59yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "60s", "AGE_GROUP"] = "60-69yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "70s", "AGE_GROUP"] = "70-79yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "80s", "AGE_GROUP"] = "80-89yrs"
case_report_df.loc[case_report_df["AGE_GROUP"] == "90+", "AGE_GROUP"] = "90yrs+"
# Gender
gender_list = case_report_df['GENDER']
case_report_df['GENDER'] = [gender.lower().capitalize() for gender in gender_list]
# CaseAcquisition_Info
case_info_list = case_report_df['CASEACQUISITION_INFO']
case_report_df['CASEACQUISITION_INFO'] = [string.capwords(case_info.lower()) for case_info in case_info_list]
case_report_df.loc[case_report_df["CASEACQUISITION_INFO"] == "Cc", "CASEACQUISITION_INFO"] = "Close Contact"
case_report_df.loc[case_report_df["CASEACQUISITION_INFO"] == "Ob", "CASEACQUISITION_INFO"] = "Outbreak"
case_report_df.loc[case_report_df["CASEACQUISITION_INFO"] == "Travel", "CASEACQUISITION_INFO"] = "Travel-Related"

# Chuyển đổi kiểu dữ liệu
case_report_df['SPECIMENDATE'] = pd.to_datetime(case_report_df.SPECIMENDATE)
case_report_df['REPORTED_DATE'] = pd.to_datetime(case_report_df.REPORTED_DATE)
case_report_df['TESTREPORTED_DATE'] = pd.to_datetime(case_report_df.TESTREPORTED_DATE)
case_report_df['ACCURATEEPISODE_DT'] = pd.to_datetime(case_report_df.ACCURATEEPISODE_DT)

# Xem dữ liệu
case_report_df.head()

OUTCOME AGE_GROUP  GENDER                      REPORTING_PHU SPECIMENDATE  \
0  Resolved  80-89yrs  Female              Toronto Public Health   2020-09-24   
1  Resolved    <20yrs    Male                 Peel Public Health   2020-12-24   
2  Resolved  20-29yrs    Male              Lambton Public Health   2020-09-23   
3  Resolved  20-29yrs    Male  Region of Waterloo, Public Health   2020-09-20   
4  Resolved  20-29yrs  Female               Ottawa Public Health   2020-09-21   

  REPORTED_DATE REPORTING_PHU_CITY TESTREPORTED_DATE CASEACQUISITION_INFO  \
0    2021-02-06            Toronto        2021-02-06             Outbreak   
1    2021-02-01        Mississauga        2021-02-01        Close Contact   
2    2021-01-15       Point Edward        2021-01-14        Close Contact   
3    2020-12-29           Waterloo        2020-09-23    No Known Epi Link   
4    2020-12-24             Ottawa        2020-09-22        Close Contact   

  ACCURATEEPISODE_DT           REPORTING_PHU_ADDRESS  \
0         2021-01-26  277 Victoria Street, 5th Floor   
1         2021-01-29          7120 Hurontario Street   
2         2021-01-10              160 Exmouth Street   
3         2020-09-20          99 Regina Street South   
4         2020-09-21         100 Constellation Drive   

                               REPORTING_PHU_WEBSITE OUTBREAKRELATED  \
0  www.toronto.ca/community-people/health-wellnes...             Yes   
1                          www.peelregion.ca/health/             NaN   
2                            www.lambtonhealth.on.ca             Yes   
3                            www.regionofwaterloo.ca             NaN   
4                          www.ottawapublichealth.ca             NaN   

   REPORTING_PHU_LATITUDE  REPORTING_PHU_LONGITUDE REPORTING_PHU_POSTAL_CODE  
0               43.656591               -79.379358                   M5B 1W2  
1               43.647471               -79.708893                   L5W 1N4  
2               42.986416               -82.404808                   N7T 7Z6  
3               43.462876               -80.520913                   N2J 4V3  
4               45.345665               -75.763912                   K2G 6J8

### Xem thông tin số dòng và các cột của bảng

In [10]:
case_report_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28500 entries, 0 to 28499
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   OUTCOME                    28500 non-null  object        
 1   AGE_GROUP                  28500 non-null  object        
 2   GENDER                     28500 non-null  object        
 3   REPORTING_PHU              28500 non-null  object        
 4   SPECIMENDATE               28292 non-null  datetime64[ns]
 5   REPORTED_DATE              28500 non-null  datetime64[ns]
 6   REPORTING_PHU_CITY         28500 non-null  object        
 7   TESTREPORTED_DATE          28082 non-null  datetime64[ns]
 8   CASEACQUISITION_INFO       28500 non-null  object        
 9   ACCURATEEPISODE_DT         28500 non-null  datetime64[ns]
 10  REPORTING_PHU_ADDRESS      28500 non-null  object        
 11  REPORTING_PHU_WEBSITE      28500 non-null  object        
 12  OUTB

$\rightarrow$ **Kết quả:** Có tổng cộng 28500 dòng dữ liệu. Các cột `SpecimenDate`, `TestReportedDate` và `OutbreakRelated` còn khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [11]:
# Kiểm tra các dòng bị trùng lắp
case_report_df[case_report_df.duplicated()]

OUTCOME AGE_GROUP  GENDER                       REPORTING_PHU  \
105    Resolved    <20yrs  Female     Hamilton Public Health Services   
115    Resolved  20-29yrs  Female                  Peel Public Health   
134    Resolved  20-29yrs    Male                  Peel Public Health   
162    Resolved  30-39yrs  Female                  Peel Public Health   
197    Resolved    <20yrs  Female     Hamilton Public Health Services   
...         ...       ...     ...                                 ...   
27070  Resolved  30-39yrs  Female                  Peel Public Health   
27102  Resolved  40-49yrs    Male  York Region Public Health Services   
27509  Resolved  20-29yrs  Female                  Peel Public Health   
27720  Resolved  60-69yrs  Female                  Peel Public Health   
27786  Resolved  30-39yrs    Male                  Peel Public Health   

      SPECIMENDATE REPORTED_DATE REPORTING_PHU_CITY TESTREPORTED_DATE  \
105     2020-10-06    2020-10-09           Hamilton        2020-10-09   
115     2020-10-03    2020-10-09        Mississauga        2020-10-09   
134     2020-10-03    2020-10-09        Mississauga        2020-10-09   
162     2020-10-07    2020-10-09        Mississauga        2020-10-09   
197     2020-10-06    2020-10-08           Hamilton        2020-10-08   
...            ...           ...                ...               ...   
27070   2020-03-28    2020-03-30        Mississauga        2020-03-30   
27102   2020-03-26    2020-03-30          Newmarket        2020-03-30   
27509   2020-03-25    2020-03-28        Mississauga        2020-03-28   
27720   2020-03-21    2020-03-27        Mississauga        2020-03-27   
27786   2020-03-20    2020-03-26        Mississauga        2020-03-26   

      CASEACQUISITION_INFO ACCURATEEPISODE_DT         REPORTING_PHU_ADDRESS  \
105          Close Contact         2020-10-06  110 King St. West, 2nd Floor   
115          Close Contact         2020-10-03        7120 Hurontario Street   
134          Close Contact         2020-10-03        7120 Hurontario Street   
162          Close Contact         2020-10-07        7120 Hurontario Street   
197               Outbreak         2020-10-04  110 King St. West, 2nd Floor   
...                    ...                ...                           ...   
27070        Close Contact         2020-03-23        7120 Hurontario Street   
27102    No Known Epi Link         2020-03-19            17250 Yonge Street   
27509        Close Contact         2020-03-24        7120 Hurontario Street   
27720        Close Contact         2020-03-14        7120 Hurontario Street   
27786    No Known Epi Link         2020-03-17        7120 Hurontario Street   

                          REPORTING_PHU_WEBSITE OUTBREAKRELATED  \
105                www.hamilton.ca/publichealth             NaN   
115                   www.peelregion.ca/health/             NaN   
134                   www.peelregion.ca/health/             NaN   
162                   www.peelregion.ca/health/             NaN   
197                www.hamilton.ca/publichealth             Yes   
...                                         ...             ...   
27070                 www.peelregion.ca/health/             NaN   
27102  www.york.ca/wps/portal/yorkhome/health/              NaN   
27509                 www.peelregion.ca/health/             NaN   
27720                 www.peelregion.ca/health/             NaN   
27786                 www.peelregion.ca/health/             NaN   

       REPORTING_PHU_LATITUDE  REPORTING_PHU_LONGITUDE  \
105                 43.257631               -79.871341   
115                 43.647471               -79.708893   
134                 43.647471               -79.708893   
162                 43.647471               -79.708893   
197                 43.257631               -79.871341   
...                       ...                      ...   
27070               43.647471               -79.708893   
27102               44.048023       

$\rightarrow$ **Kết quả:** Có tổng cộng 1528 dòng dữ liệu bị trùng lắp.

In [12]:
# Xử lí trùng lắp bằng cách xóa dòng
case_report_df = case_report_df.drop_duplicates()

In [13]:
# Kiểm tra xem các dòng bị trùng lắp đã được xóa chưa
case_report_df[case_report_df.duplicated()]

Empty DataFrame
Columns: [OUTCOME, AGE_GROUP, GENDER, REPORTING_PHU, SPECIMENDATE, REPORTED_DATE, REPORTING_PHU_CITY, TESTREPORTED_DATE, CASEACQUISITION_INFO, ACCURATEEPISODE_DT, REPORTING_PHU_ADDRESS, REPORTING_PHU_WEBSITE, OUTBREAKRELATED, REPORTING_PHU_LATITUDE, REPORTING_PHU_LONGITUDE, REPORTING_PHU_POSTAL_CODE]
Index: []

$\rightarrow$ **Kết quả:** Bảng `Cases Report` đã hết các dòng dữ liệu bị trùng lắp.

### Xuất dữ liệu hoàn chỉnh ra file source

In [14]:
case_report_df.to_csv('Datasource_Cleaned/Case_Report.csv', sep=';')

## 2. Chuẩn bị dữ liệu `Compiled_COVID-19_Case_Details_(Canada)`

### Đọc và xem bảng

In [15]:
# Đọc dữ liệu
f_path = 'Datasource/Compiled_COVID-19_Case_Details_(Canada).csv'
case_detail_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = case_detail_df.columns.str.title()
case_detail_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
case_detail_df.head()

OBJECTID  ROW_ID           DATE_REPORTED  \
0         1     500  2020/03/23 12:00:00+00   
1         2     501  2020/04/02 12:00:00+00   
2         3     502  2020/03/25 12:00:00+00   
3         4     503  2020/03/28 12:00:00+00   
4         5     504  2020/03/30 12:00:00+00   

                                       HEALTH_REGION AGE_GROUP  GENDER  \
0                                 Peel Public Health     20-29  Female   
1                                 Peel Public Health     40-49  Female   
2  Kingston, Frontenac and Lennox & Addington Pub...     30-39    Male   
3  Kingston, Frontenac and Lennox & Addington Pub...       <20  Female   
4                                 Peel Public Health     50-59    Male   

        EXPOSURE CASE_STATUS PROVINCE  
0  Close Contact   Recovered  Ontario  
1  Close Contact   Recovered  Ontario  
2  Close Contact   Recovered  Ontario  
3  Close Contact   Recovered  Ontario  
4  Close Contact   Recovered  Ontario

### Matching Data

In [16]:
# Đổi tên một số cột theo bảng mô tả dữ liệu của để bài
case_detail_df.rename(columns = {'DATE_REPORTED':'REPORTED_DATE', 'HEALTH_REGION':'REPORTING_PHU', 'PHUCITY':'REPORTING_PHU_CITY', 'PHU_ADDRESS':'REPORTING_PHU_ADDRESS', 'PHU_WEBSITE':'REPORTING_PHU_WEBSITE', 'PHU_LATITUDE':'REPORTING_PHU_LATITUDE', 'PHU_LONGITUDE':'REPORTING_PHU_LONGITUDE', 'PHU_POSTAL_CODE':'REPORTING_PHU_POSTAL_CODE'}, inplace = True)

# Đổi value của một số cột để khớp với các bảng
# Age_Group
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "<20", "AGE_GROUP"] = "<20yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "20-29", "AGE_GROUP"] = "20-29yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "30-39", "AGE_GROUP"] = "30-39yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "40-49", "AGE_GROUP"] = "40-49yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "50-59", "AGE_GROUP"] = "50-59yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "60-69", "AGE_GROUP"] = "60-69yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "70-79", "AGE_GROUP"] = "70-79yrs"
case_detail_df.loc[case_detail_df["AGE_GROUP"] == "80+", "AGE_GROUP"] = "80yrs+"

# Chuyển đổi kiểu dữ liệu
case_detail_df['REPORTED_DATE'] = pd.to_datetime(case_detail_df.REPORTED_DATE)

# Xem dữ liệu
case_detail_df.head()

OBJECTID  ROW_ID             REPORTED_DATE  \
0         1     500 2020-03-23 12:00:00+00:00   
1         2     501 2020-04-02 12:00:00+00:00   
2         3     502 2020-03-25 12:00:00+00:00   
3         4     503 2020-03-28 12:00:00+00:00   
4         5     504 2020-03-30 12:00:00+00:00   

                                       REPORTING_PHU AGE_GROUP  GENDER  \
0                                 Peel Public Health  20-29yrs  Female   
1                                 Peel Public Health  40-49yrs  Female   
2  Kingston, Frontenac and Lennox & Addington Pub...  30-39yrs    Male   
3  Kingston, Frontenac and Lennox & Addington Pub...    <20yrs  Female   
4                                 Peel Public Health  50-59yrs    Male   

        EXPOSURE CASE_STATUS PROVINCE  
0  Close Contact   Recovered  Ontario  
1  Close Contact   Recovered  Ontario  
2  Close Contact   Recovered  Ontario  
3  Close Contact   Recovered  Ontario  
4  Close Contact   Recovered  Ontario

### Xem thông tin số dòng và các cột của bảng

In [17]:
case_detail_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype              
---  ------         --------------    -----              
 0   OBJECTID       1048575 non-null  int64              
 1   ROW_ID         1048575 non-null  int64              
 2   REPORTED_DATE  1048575 non-null  datetime64[ns, UTC]
 3   REPORTING_PHU  1048575 non-null  object             
 4   AGE_GROUP      1048575 non-null  object             
 5   GENDER         1048575 non-null  object             
 6   EXPOSURE       1048575 non-null  object             
 7   CASE_STATUS    1048575 non-null  object             
 8   PROVINCE       1048575 non-null  object             
dtypes: datetime64[ns, UTC](1), int64(2), object(6)
memory usage: 72.0+ MB


$\rightarrow$ **Kết quả:** Có tổng cộng 1048575 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [18]:
# Kiểm tra các dòng bị trùng lắp
case_detail_df[case_detail_df.duplicated()]

Empty DataFrame
Columns: [OBJECTID, ROW_ID, REPORTED_DATE, REPORTING_PHU, AGE_GROUP, GENDER, EXPOSURE, CASE_STATUS, PROVINCE]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [19]:
case_detail_df.to_csv('Datasource_Cleaned/Case_Detail.csv', sep=';')

## 3. Chuẩn bị dữ liệu `ongoing_outbreaks_phu`

### Đọc và xem bảng

In [20]:
# Đọc dữ liệu
f_path = 'Datasource/ongoing_outbreaks_phu.csv'
ongoing_outbreak_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = ongoing_outbreak_df.columns.str.title()
ongoing_outbreak_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
ongoing_outbreak_df.head()

DATE  PHU_NUM     OUTBREAK_GROUP  NUMBER_ONGOING_OUTBREAKS
0  11/1/2020     2227  1 Congregate Care                         2
1  11/1/2020     2227        4 Workplace                         1
2  11/1/2020     2240        4 Workplace                         1
3  11/1/2020     2240     5 Recreational                         1
4  11/1/2020     2240    6 Other/Unknown                         1

### Matching data

In [21]:
# Đổi tên một số cột theo bảng mô tả dữ liệu của để bài
ongoing_outbreak_df.rename(columns = {'DATE':'REPORTED_DATE', 'PHU_NUM':'PHU_ID'}, inplace = True)

# Chuyển đổi kiểu dữ liệu
ongoing_outbreak_df['REPORTED_DATE'] = pd.to_datetime(ongoing_outbreak_df.REPORTED_DATE)

# Xem dữ liệu
ongoing_outbreak_df.head()

REPORTED_DATE  PHU_ID     OUTBREAK_GROUP  NUMBER_ONGOING_OUTBREAKS
0    2020-11-01    2227  1 Congregate Care                         2
1    2020-11-01    2227        4 Workplace                         1
2    2020-11-01    2240        4 Workplace                         1
3    2020-11-01    2240     5 Recreational                         1
4    2020-11-01    2240    6 Other/Unknown                         1

### Xem thông tin số dòng và các cột của bảng

In [22]:
ongoing_outbreak_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56987 entries, 0 to 56986
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   REPORTED_DATE             56987 non-null  datetime64[ns]
 1   PHU_ID                    56987 non-null  int64         
 2   OUTBREAK_GROUP            56987 non-null  object        
 3   NUMBER_ONGOING_OUTBREAKS  56987 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 1.7+ MB


$\rightarrow$ **Kết quả:** Có tổng cộng 56987 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [23]:
# Kiểm tra các dòng bị trùng lắp
ongoing_outbreak_df[ongoing_outbreak_df.duplicated()]

Empty DataFrame
Columns: [REPORTED_DATE, PHU_ID, OUTBREAK_GROUP, NUMBER_ONGOING_OUTBREAKS]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [24]:
ongoing_outbreak_df.to_csv('Datasource_Cleaned/Ongoing_Outbreak.csv', sep=';')

## 4. Chuẩn bị dữ liệu `Public health unit`

### Đọc và xem bảng

In [25]:
# Đọc dữ liệu
f_path = 'Datasource/Public health unit.xlsx'
phu_df = pd.read_excel(f_path)

# Chuẩn hóa tên cột
header_list = phu_df.columns.str.title()
phu_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
phu_df.head()

PHU_ID                             REPORTING_PHU  \
0    2244              Middlesex-London Health Unit   
1    2236           Halton Region Health Department   
2    3895                     Toronto Public Health   
3    5183          Huron Perth District Health Unit   
4    2266  Wellington-Dufferin-Guelph Public Health   

            REPORTING_PHU_ADDRESS REPORTING_PHU_CITY  \
0                  50 King Street             London   
1                1151 Bronte Road           Oakville   
2  277 Victoria Street, 5th Floor            Toronto   
3            653 West Gore Street          Stratford   
4             160 Chancellors Way             Guelph   

  REPORTING_PHU_POSTAL_CODE  \
0                   N6A 5L7   
1                   L6M 3Ll   
2                   M5B 1W2   
3                   N5A 1L4   
4                   N1G 0E1   

                               REPORTING_PHU_WEBSITE  REPORTING_PHU_LATITUDE  \
0                                 www.healthunit.com              4299571559   
1         www.halton.ca/For-Residents/Public-Health/              4351209709   
2  www.toronto.ca/community-people/health-wellnes...              4372584482   
3                                        www.hpph.ca              4360051407   
4                             www.wdgpublichealth.ca              4385364154   

   REPORTING_PHU_LONGITUDE  
0               -814569832  
1              -7989050809  
2              -7938955775  
3              -8128148453  
4              -8037208498

### Xem thông tin số dòng và các cột của bảng

In [26]:
phu_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   PHU_ID                     35 non-null     int64 
 1   REPORTING_PHU              35 non-null     object
 2   REPORTING_PHU_ADDRESS      35 non-null     object
 3   REPORTING_PHU_CITY         35 non-null     object
 4   REPORTING_PHU_POSTAL_CODE  35 non-null     object
 5   REPORTING_PHU_WEBSITE      35 non-null     object
 6   REPORTING_PHU_LATITUDE     35 non-null     int64 
 7   REPORTING_PHU_LONGITUDE    35 non-null     int64 
dtypes: int64(3), object(5)
memory usage: 2.3+ KB


$\rightarrow$ **Kết quả:** Có tổng cộng 35 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [27]:
# Kiểm tra các dòng bị trùng lắp
phu_df[phu_df.duplicated()]

PHU_ID                 REPORTING_PHU REPORTING_PHU_ADDRESS  \
34    2244  Middlesex-London Health Unit        50 King Street   

   REPORTING_PHU_CITY REPORTING_PHU_POSTAL_CODE REPORTING_PHU_WEBSITE  \
34             London                   N6A 5L7    www.healthunit.com   

    REPORTING_PHU_LATITUDE  REPORTING_PHU_LONGITUDE  
34              4299571559               -814569832

$\rightarrow$ **Kết quả:** Có tổng cộng 1 dòng dữ liệu bị trùng lắp.

In [28]:
# Xử lí trùng lắp bằng cách xóa dòng
phu_df = phu_df.drop_duplicates()

In [29]:
# Kiểm tra xem các dòng bị trùng lắp đã được xóa chưa
phu_df[phu_df.duplicated()]

Empty DataFrame
Columns: [PHU_ID, REPORTING_PHU, REPORTING_PHU_ADDRESS, REPORTING_PHU_CITY, REPORTING_PHU_POSTAL_CODE, REPORTING_PHU_WEBSITE, REPORTING_PHU_LATITUDE, REPORTING_PHU_LONGITUDE]
Index: []

$\rightarrow$ **Kết quả:** Bảng `Public Health Unit` đã hết các dòng dữ liệu bị trùng lắp.

### Xuất dữ liệu hoàn chỉnh ra file source

In [30]:
phu_df.to_csv('Datasource_Cleaned/Public_Health_Unit.csv', sep=';')

## 5. Chuẩn bị dữ liệu `Public Health Units GROUP`

### Đọc và xem bảng

In [31]:
# Đọc dữ liệu
f_path = 'Datasource/Public Health Units GROUP.xlsx'
phu_group_df = pd.read_excel(f_path)

# Chuẩn hóa tên cột
header_list = phu_group_df.columns.str.title()
phu_group_df.columns = [header.upper() for header in header_list]

# Xem thử dữ liệu
phu_group_df.head()

PHU_GROUP       PHU_CITY  \
0       Central        Toronto   
1       Central      Newmarket   
2  Central East  Durham Region   
3  Central East      Port Hope   
4  Central East   Peterborough   

                                          PHU_REGION  
0                              Toronto Public Health  
1                 York Region Public Health Services  
2                    Durham Region Health Department  
3  Haliburton, Kawartha, Pine Ridge District Heal...  
4                         Peterborough Public Health

### Matching Data

In [32]:
# Đổi tên một số cột theo bảng mô tả dữ liệu của để bài
phu_group_df.rename(columns = {'PHU_GROUP':'REPORTING_PHU_GROUP', 'PHU_CITY':'REPORTING_PHU_CITY', 'PHU_REGION':'REPORTING_PHU'}, inplace = True)

# Xem dữ liệu
phu_group_df.head()

REPORTING_PHU_GROUP REPORTING_PHU_CITY  \
0             Central            Toronto   
1             Central          Newmarket   
2        Central East      Durham Region   
3        Central East          Port Hope   
4        Central East       Peterborough   

                                       REPORTING_PHU  
0                              Toronto Public Health  
1                 York Region Public Health Services  
2                    Durham Region Health Department  
3  Haliburton, Kawartha, Pine Ridge District Heal...  
4                         Peterborough Public Health

### Xem thông tin số dòng và các cột của bảng

In [33]:
phu_group_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   REPORTING_PHU_GROUP  34 non-null     object
 1   REPORTING_PHU_CITY   34 non-null     object
 2   REPORTING_PHU        34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


$\rightarrow$ **Kết quả:** Có tổng cộng 34 dòng dữ liệu và không có cột nào bị khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [34]:
# Kiểm tra các dòng bị trùng lắp
phu_group_df[phu_group_df.duplicated()]

Empty DataFrame
Columns: [REPORTING_PHU_GROUP, REPORTING_PHU_CITY, REPORTING_PHU]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [35]:
phu_group_df.to_csv('Datasource_Cleaned/Public_Health_Units_GROUP.csv', sep=';')

## 6. Chuẩn bị dữ liệu `vaccines_by_age_phu`

### Đọc và xem bảng

In [36]:
# Đọc dữ liệu
f_path = 'Datasource/vaccines_by_age_phu.csv'
vaccines_df = pd.read_csv(f_path)

# Chuẩn hóa tên cột
header_list = vaccines_df.columns.str.title()
vaccines_df.columns = [header.replace(' ','_').upper() for header in header_list]

# Xem thử dữ liệu
vaccines_df.head()

DATE  PHU_ID  AGEGROUP  AT_LEAST_ONE_DOSE_CUMULATIVE  \
0  7/26/2021    2226  12-17yrs                        3843.0   
1  7/26/2021    2227  12-17yrs                        6306.0   
2  7/26/2021    2230  12-17yrs                       34025.0   
3  7/26/2021    2233  12-17yrs                        5533.0   
4  7/26/2021    2234  12-17yrs                        3125.0   

   SECOND_DOSE_CUMULATIVE  FULLY_VACCINATED_CUMULATIVE  THIRD_DOSE_CUMULATIVE  
0                  1890.0                          NaN                    NaN  
1                  3567.0                          NaN                    NaN  
2                 16953.0                          NaN                    NaN  
3                  3627.0                          NaN                    NaN  
4                  1711.0                          NaN                    NaN

### Matching data

In [37]:
# Đổi tên một số cột theo bảng mô tả dữ liệu của để bài
vaccines_df.rename(columns = {'DATE':'REPORTED_DATE', 'AGEGROUP':'AGE_GROUP'}, inplace = True)

# Chuyển đổi kiểu dữ liệu
vaccines_df['REPORTED_DATE'] = pd.to_datetime(vaccines_df.REPORTED_DATE)
vaccines_df.loc[vaccines_df["AGE_GROUP"] == "80+", "AGE_GROUP"] = "80yrs+"

# Xem dữ liệu
vaccines_df.head()

REPORTED_DATE  PHU_ID AGE_GROUP  AT_LEAST_ONE_DOSE_CUMULATIVE  \
0    2021-07-26    2226  12-17yrs                        3843.0   
1    2021-07-26    2227  12-17yrs                        6306.0   
2    2021-07-26    2230  12-17yrs                       34025.0   
3    2021-07-26    2233  12-17yrs                        5533.0   
4    2021-07-26    2234  12-17yrs                        3125.0   

   SECOND_DOSE_CUMULATIVE  FULLY_VACCINATED_CUMULATIVE  THIRD_DOSE_CUMULATIVE  
0                  1890.0                          NaN                    NaN  
1                  3567.0                          NaN                    NaN  
2                 16953.0                          NaN                    NaN  
3                  3627.0                          NaN                    NaN  
4                  1711.0                          NaN                    NaN

### Xem thông tin số dòng và các cột của bảng

In [38]:
vaccines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 171182 entries, 0 to 171181
Data columns (total 7 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   REPORTED_DATE                 171182 non-null  datetime64[ns]
 1   PHU_ID                        171182 non-null  int64         
 2   AGE_GROUP                     171182 non-null  object        
 3   AT_LEAST_ONE_DOSE_CUMULATIVE  170971 non-null  float64       
 4   SECOND_DOSE_CUMULATIVE        48642 non-null   float64       
 5   FULLY_VACCINATED_CUMULATIVE   121912 non-null  float64       
 6   THIRD_DOSE_CUMULATIVE         106015 non-null  float64       
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 9.1+ MB


$\rightarrow$ **Kết quả:** Có tổng cộng 171182 dòng dữ liệu. Các cột `At least one dose_cumulative`, `Second_dose_cumulative`, `fully_vaccinated_cumulative` và `third_dose_cumulative` còn khuyết dữ liệu.

### Kiểm tra dữ liệu bị duplicated và xử lí

In [39]:
# Kiểm tra các dòng bị trùng lắp
vaccines_df[vaccines_df.duplicated()]

Empty DataFrame
Columns: [REPORTED_DATE, PHU_ID, AGE_GROUP, AT_LEAST_ONE_DOSE_CUMULATIVE, SECOND_DOSE_CUMULATIVE, FULLY_VACCINATED_CUMULATIVE, THIRD_DOSE_CUMULATIVE]
Index: []

$\rightarrow$ **Kết quả:** Không có dòng dữ liệu bị duplicated nên không cần xử lí.

### Xuất dữ liệu hoàn chỉnh ra file source

In [40]:
vaccines_df.to_csv('Datasource_Cleaned/Vaccines.csv', sep=';')